File: src/main/java/com/github/ccguyka/showupdates/slack/SlackClient.java
package com.github.ccguyka.showupdates.slack;

import java.io.IOException;
import java.util.Arrays;

import com.github.ccguyka.showupdates.objects.ProjectUpdates;
import com.github.seratch.jslack.Slack;
import com.github.seratch.jslack.api.model.block.LayoutBlock;
import com.github.seratch.jslack.api.model.block.SectionBlock;
import com.github.seratch.jslack.api.model.block.composition.MarkdownTextObject;
import com.github.seratch.jslack.api.webhook.Payload;
import com.github.seratch.jslack.api.webhook.WebhookResponse;

public class SlackClient {

    private ProjectUpdates projectUpdates;
    private String token;
    private String artifact;

    public SlackClient(String artifact) {
        this.artifact = artifact;
    }

    public static SlackClient forArtifact(String artifact) {
        return new SlackClient(artifact);
    }

    public SlackClient updates(ProjectUpdates projectUpdates) {
        this.projectUpdates = projectUpdates;

        return this;
    }

    public SlackClient token(String token) {
        this.token = token;

        return this;
    }

    public Response send() throws IOException {
        String markdownText = SlackBody.artifact(artifact)
                .updates(projectUpdates)
                .build();
        LayoutBlock block = SectionBlock.builder()
                .text(MarkdownTextObject.builder()
                        .text(markdownText)
                        .build())
                .build();
        Payload payload = Payload.builder()
                .blocks(Arrays.asList(block))
                .build();

        Slack slack = Slack.getInstance();
        return Response.response(slack.send(token, payload));
    }

    public static class Response {

        private Integer code;
        private String message;

        private Response(Integer code, String message) {
            this.code = code;
            this.message = message;
        }

        static Response response(WebhookResponse response) {
            return new Response(response.getCode(), response.getBody());
        }

        public Integer getCode() {
            return code;
        }

        public String getMessage() {
            return message;
        }
    }

}

File: src/main/kotlin/com/github/ccguyka/showupdates/slack/SlackClient.kt
package com.github.ccguyka.showupdates.slack

import com.github.ccguyka.showupdates.objects.ProjectUpdates
import com.github.ccguyka.showupdates.slack.SlackBody.Companion.artifact
import com.slack.api.Slack
import com.slack.api.model.block.LayoutBlock
import com.slack.api.model.block.SectionBlock
import com.slack.api.model.block.composition.MarkdownTextObject
import com.slack.api.webhook.Payload
import com.slack.api.webhook.WebhookResponse
import java.io.IOException

class SlackClient(private val artifact: String?) {

    private var projectUpdates: ProjectUpdates? = null
    private var token: String? = null

    fun updates(projectUpdates: ProjectUpdates?): SlackClient {
        this.projectUpdates = projectUpdates
        return this
    }

    fun token(token: String?): SlackClient {
        this.token = token
        return this
    }

    @Throws(IOException::class)
    fun send(): Response {
        val markdownText = artifact(artifact)
                .updates(projectUpdates)
                .build()
        val block: LayoutBlock = SectionBlock.builder()
                .text(MarkdownTextObject.builder()
                        .text(markdownText)
                        .build())
                .build()
        val payload = Payload.builder()
                .blocks(listOf(block))
                .build()
        val slack = Slack.getInstance()
        return Response.response(slack.send(token, payload))
    }

    class Response private constructor(val code: Int, val message: String) {

        companion object {
            fun response(response: WebhookResponse): Response {
                return Response(response.code, response.body)
            }
        }
    }

    companion object {
        @JvmStatic
        fun forArtifact(artifact: String?): SlackClient {
            return SlackClient(artifact)
        }
    }
}
